### ISSUE TITLE
hitting max path component limit

### ISSUE THREAD

**Issue by jhgit** [ninja-build/ninja#1161]
I hit this error on a non-windows (bsd) system:

ninja: fatal: path has too many components

My first question upon hitting that is - what is meant by "too many"?  Why did I hit this?

Opening the source, I see the code in CanonicalizePath in src/util.cc - so I have the answer to what is meant by "how many".  Turns out, I'm building llvm in a mount point.  The project already adds a few directory levels.  Then on top of that, my mount paths added more.  After all was said and done, I had a path with 32 components which exceeds the hard-coded kMaxPathComponents (30).

Okay, so now what?
- Accept the limitation and work in a shallower hierarchy.  Maybe it could be documented better, but that's fine for now.  I'm sure 99% of ninja uses won't hit this problem.
- But can this be improved in ninha?
  - better error mesage? - sure, easy enough
  - extend the hard-coded limit?  A little fragile due to the current algorithm (thanks a lot, Windows), but one could modify that (64?) and still retain decent performance.
  - gracefully transition to a less performant mode when long paths are encountered?  Maybe add a footnote in the docs about avoiding shooting oneself in the foot, performance-wise?

I'm willing to concoct a patch and do some benchmarking, but I'd like to know the project's feeling about this.  Where is the thinking on the spectrum from "hell, no - if you have a path that long, you're an idiot - we're not changing anything" to "yeah, we could document that a little better" to "sure, if the OS can support a longer path, why not modify ninja to allow that"?

I also haven't done the profiling/inspection necessary to understand how often paths need to be canonicalized (and whether it's necessary in all cases where it's used - i.e., why bother spending cycles canonicalizing at all?).

**Comment by martine** [comment]
Thanks for researching the solutions!  You're the first person to ever run into this problem, so it's a bit of a surprise.  The intent of the error message was mostly to direct someone at the bug tracker, since we didn't expect to hit it -- I definitely should've made it clearer.

My first guess is that just upping the limit would be the easiest solution.  The only reason this is a fixed value at all is that we didn't want to spend extra time growing an array.  Unfortunately, from a glance at the code it looks like we use a 32-bit bitfield for some windows backslash trickery.  Nico can likely comment on whether making it a 64-bit bitfield would matter.

Path canonicalization has historically been an important bottleneck in Ninja, which might come as a surprise.  But Ninja doesn't do a whole lot other than load up the .ninja files and stat a bunch of stuff, so it spends most of the work it does do on munging paths.  We have to canonicalize to ensure we always use the same path to refer to the same file, and we can't define away the problem because our input data can come from the output of `gcc -MD` which may have joined together multiple variables to create a funky path.  There's also historically been lots of people who have trouble on Windows convincing all the tools to use either forward or backward slashes so we end up needing to support both.

**Comment by nico** [comment]
If using 64-bit ints instead of 32-bit ints for that bitmask doesn't regress performance, that sounds fine to me.

**Comment by ArjanKn** [comment]
OK so i'm not the only one running into this issues. I'm on windows btw. Did no see a pr with the suggested increased size, is it in already?

**Comment by gunrot** [comment]
Today i was also hit by this bug. I use qt and make out of source build and moc generates #include with lot of .. 
````
I patched it like this to make it work:
diff --git a/src/util.cc b/src/util.cc
index e31fd1f..0848824 100644
--- a/src/util.cc
+++ b/src/util.cc
@@ -125,7 +125,7 @@ bool CanonicalizePath(char* path, size_t* len, unsigned int* slash_bits,
     return false;
   }

-  const int kMaxPathComponents = 30;
+  const int kMaxPathComponents = 60;
   char* components[kMaxPathComponents];
   int component_count = 0;

@@ -136,7 +136,7 @@ bool CanonicalizePath(char* path, size_t* len, unsigned int* slash_bits,

 #ifdef _WIN32
   unsigned int bits = 0;
-  unsigned int bits_mask = 1;
+  unsigned long int bits_mask = 1;
   int bits_offset = 0;
   // Convert \ to /, setting a bit in |bits| for each \ encountered.
   for (char* c = path; c < end; ++c) {
@@ -150,7 +150,7 @@ bool CanonicalizePath(char* path, size_t* len, unsigned int* slash_bits,
         bits_offset++;
     }
   }
-  if (bits_offset > 32) {
+  if (bits_offset > 64) {
     *err = "too many path components";
     return false;
   }
````

**Comment by DanielWeber** [comment]
I opened PR 1181 to increase the limits, based on the patch that gunrot pasted above. But with some adaptations.

**Comment by randomascii** [comment]
We just hit this on a Chromium build:

https://luci-logdog.appspot.com/v/?s=chromium%2Fbb%2Ftryserver.chromium.win%2Fwin_chromium_rel_ng%2F482153%2F%2B%2Frecipes%2Fsteps%2Fcompile__with_patch_%2F0%2Fstdout

**Comment by danjacques** [comment]
Also hit this in another Chromium build. The build environment here has a slightly deeper path nesting than standard Chromium:

https://luci-logdog.appspot.com/v/?s=chromium%2Fbuildbucket%2Fcr-buildbucket.appspot.com%2F8974709182564061120%2F%2B%2Fsteps%2Fcompile__with_patch_%2F0%2Fstdout

**Comment by egorpugin** [comment]
Also hit this not in Chromium build, but in my custom tool build.
So +1 from me.

**Comment by nico** [comment]
We fixed this in #1181, just need to do a release with that fix.
